defender/staged/downloaded/Compound_Gasless_Voting.js

```diff
@@ -0,0 +1,95 @@
+];
+
+exports.handler = async function handler(autotaskEvent) {
+  if (autotaskEvent.request === undefined) {
+    throw new Error('POST request information not found');
+  }
+  // body contains batch of signed messages to submit on-chain through the Defender Relayer
+  // the api endpoint does most of the validation checking for us already
+  // (ex. checks for valid signatures and all required fields)
+  // https://app.swaggerhub.com/apis-docs/arr00/COMP.vote/1.0#/
+  const { body } = autotaskEvent.request;
+
+  if (Array.isArray(body) === false) {
+    throw new Error('Request body must be an Array');
+  }
+
+  // initiate a new relay provider
+  console.debug('Creating DefenderRelayProvider');
+  const provider = new DefenderRelayProvider(autotaskEvent);
+
+  // initiate a new relay signer
+  console.debug('Creating DefenderRelaySigner');
+  const signer = new DefenderRelaySigner(autotaskEvent, provider, { speed: 'fast' });
+
+  // create an ethers.js Contract Object to interact with the on-chain governance smart contract
+  console.debug('Creating governanceContract');
+  const governanceContract = new ethers.Contract(
+    compoundGovernanceAddress,
+    compoundGovernanceAbi,
+    signer,
+  );
+
+  // create an ethers.js Contract Object to interact with the on-chain COMP smart contract
+  console.debug('Creating compContract');
+  const compTokenContract = new ethers.Contract(
+    compTokenAddress,
+    compTokenAbi,
+    signer,
+  );
+
+  // cast vote or delegate vote on-chain
+  const promises = body.map(async (votes) => {
+    if (votes.delegatee !== undefined) {
+      const {
+        address, delegatee, nonce, expiry, v, r, s,
+      } = votes;
+
+      // delegate vote on-chain
+      console.debug(`Delgating votes from ${address} to ${delegatee}`);
+      try {
+        await compTokenContract.delegateBySig(delegatee, nonce, expiry, v, r, s);
+      } catch (error) {
+        console.error(error);
+      }
+    } else if (votes.support !== undefined) {
+      const {
+        address, proposalId, support, v, r, s,
+      } = votes;
+      if (support === 0) {
+        console.debug(`Address ${address} is casting a vote in favor of proposal ID: ${proposalId}`);
+      } else if (support === 1) {
+        console.debug(`Address ${address} is casting a vote against proposal ID: ${proposalId}`);
+      } else if (support === 2) {
+        console.debug(`Address ${address} is casting a vote to abstain for proposal ID: ${proposalId}`);
+      }
+
+      // cast vote on-chain
+      try {
+        await governanceContract.castVoteBySig(proposalId, support, v, r, s);
+      } catch (error) {
+        console.error(error);
+      }
+    } else {
+      // error is not thrown to prevent one failed transaction from failing the rest in the batch
+      console.error('Action is not a vote or delgate-vote');
+    }
+  });
+  await Promise.all(promises);
```

Comment:
What happens to these votes if an error occurs?  Are they lost?  Is the voter notified?

Reply:
As of right now, the transaction will fail and the vote will not go through and the failed transaction message just gets logged out. It would be great to let the voter know, but I'm not too sure the best way to handle this, as we don't have any email addresses or other forms of notification channels from the users

Reply:
Yeah, I assumed the autotask had no recourse other then to throw an error.  I was mostly making sure we're throwing the correct errors.